### PR TITLE
Add debugger codelens action

### PR DESCRIPTION
**Pull request by jaisnan**
### Description of changes: 

This PR adds the debugging action to Kani using the cargo artifact for the crate under usage.

Screen Shot 2023-05-11 at 4 45 24 PM

### Testing:

* How is this change tested? Manually

* Is this a refactor change?

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/debugger/debugger.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		return undefined;
+	}
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export async function connectToDebugger(functionName: string) {
+	const binaryName = await getBinaryPath();
+	console.log(binaryName);
+
+	// /home/ubuntu/test-concrete/target/debug/deps/test_concrete-6387540054886e23 kani_concrete_playback_check_estimate_size_14615086421508420155 --exact --nocapture
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:51]: ??

File: src/debugger/debugger.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+import { execSync } from 'child_process';
+import path from 'path';
+
+import * as vscode from 'vscode';
+
+import { getPackageName } from '../model/runCargoTest';
+import { getRootDir, getRootDirURI } from '../utils';
+
+async function getBinaryPath(): Promise<string | undefined> {
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:11]: Needs a comment.

File: src/debugger/debugger.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+		// todo -> execFile this
+		const cargoTestCommand: string = `RUSTFLAGS="--cfg=kani" cargo +nightly test --no-run --message-format=json`;
+		const output = execSync(`cd ${directory} && ${cargoTestCommand}`);
+
+		const outputString = output.toString();
+
+		const lines = outputString.trim().split('\n');
+		const jsonMessages = lines.map((line: string) => JSON.parse(line));
+		const packageCompilationArtifact = jsonMessages[jsonMessages.length - 2];
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:32]: why -2?

File: src/debugger/debugger.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		// same here
+		const directory = path.resolve(getRootDir());
+		const options = {
+			shell: false,
+			cwd: directory,
+		};
+
+		// todo -> execFile this
+		const cargoTestCommand: string = `RUSTFLAGS="--cfg=kani" cargo +nightly test --no-run --message-format=json`;
+		const output = execSync(`cd ${directory} && ${cargoTestCommand}`);
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:26]: Open an issue?

File: src/debugger/debugger.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+				reason: "compiler-artifact",
+				package_id: "test-concrete 0.1.0 (path+file:///home/ubuntu/test-concrete)",
+				...
+			},
+			{
+				reason: 'build-finished',
+				success: true
+			}
+		]
+		*/
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:45]: Can you use the multi-line syntax shown in [TSDoc](https://tsdoc.org/)?

File: src/debugger/debugger.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		const executablePath = packageCompilationArtifact.executable;
+		// Find the metadata message
+		return executablePath;
+	} catch (error) {
+		console.error('Error occurred while retrieving the binary path:', error);
+		return undefined;
+	}
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:61]: But this was a warning now, right? Didn't you want to remove this?

File: src/debugger/debugger.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		name: 'test ${programName}',
+		program: binaryName,
+		args: [functionName, '--exact', '--nocapture'],
+		cwd: '${workspaceFolder}',
+		sourceLanguages: ['rust'],
+		env: {
+			RUST_BACKTRACE: 'short',
+			RUSTFLAGS: 'cfg=kani',
+		},
+		// other configuration options specific to Rust debugging with CodeLLDB
```
Comment (adpaco-aws) [inline src/debugger/debugger.ts:79]: Not sure what this comment indicates.

Comment (jaisnan) [reply src/debugger/debugger.ts:79]: I can add a clarifying comment and move it to go above the line of code.

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**